File: packages/b2c-tooling-sdk/src/operations/bm-roles/ocapi-backend.ts
/*
 * Copyright (c) 2025, Salesforce, Inc.
 * SPDX-License-Identifier: Apache-2
 * For full license text, see the license.txt file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
 */
import type {B2CInstance} from '../../instance/index.js';
import type {components as OcapiComponents} from '../../clients/ocapi.generated.js';
import type {components as ScapiComponents} from '../../clients/scapi-merchant-roles.generated.js';
import type {
  RolesBackend,
  RoleInfo,
  ListRolesResult,
  ListRolesOptions,
  RolePermissionsInfo,
  CreateRoleInput,
} from './types.js';
import type {BmRole, BmRolePermissions} from './roles.js';
import {
  listBmRoles as ocapiListBmRoles,
  getBmRole as ocapiGetBmRole,
  createBmRole as ocapiCreateBmRole,
  deleteBmRole as ocapiDeleteBmRole,
  getBmRolePermissions as ocapiGetBmRolePermissions,
  setBmRolePermissions as ocapiSetBmRolePermissions,
  grantBmRole as ocapiGrantBmRole,
  revokeBmRole as ocapiRevokeBmRole,
} from './roles.js';

function mapOcapiRole(ocapi: BmRole): RoleInfo {
  return {
    id: ocapi.id ?? '',
    description: ocapi.description,
    userCount: ocapi.user_count,
    userManager: ocapi.user_manager,
    // OCAPI returns permissions inline on the role when expanded, same as SCAPI.
    // Map snake_case → camelCase to match the canonical RoleInfo shape so that
    // callers see consistent data after a fallback from SCAPI to OCAPI.
    permissions: ocapi.permissions ? mapOcapiPermissions(ocapi.permissions) : undefined,
    _raw: ocapi,
  };
}

type OcapiModulePermission = OcapiComponents['schemas']['role_module_permission'];
type OcapiFunctionalPermission = OcapiComponents['schemas']['role_functional_permission'];
type OcapiLocalePermission = OcapiComponents['schemas']['role_locale_permission'];
type OcapiWebdavPermission = OcapiComponents['schemas']['role_webdav_permission'];
type ScapiModulePermission = ScapiComponents['schemas']['RoleModulePermission'];
type ScapiFunctionalPermission = ScapiComponents['schemas']['RoleFunctionalPermission'];
type ScapiLocalePermission = ScapiComponents['schemas']['RoleLocalePermission'];
type ScapiWebdavPermission = ScapiComponents['schemas']['RoleWebdavPermission'];

function mapOcapiModulePermission(permission: OcapiModulePermission): ScapiModulePermission {
  return {
    application: permission.application,
    name: permission.name,
    type: permission.type,
    system: permission.system,
    value: permission.value,
    values: permission.values,
  };
}

function mapOcapiFunctionalPermission(permission: OcapiFunctionalPermission): ScapiFunctionalPermission {
  return {
    name: permission.name,
    type: permission.type,
    value: permission.value,
    values: permission.values,
  };
}

function mapOcapiLocalePermission(permission: OcapiLocalePermission): ScapiLocalePermission {
  return {
    localeId: permission.locale_id,
    type: permission.type,
    value: permission.value,
    values: permission.values,
  };
}

function mapOcapiWebdavPermission(permission: OcapiWebdavPermission): ScapiWebdavPermission {
  return {
    folder: permission.folder,
    type: permission.type,
    value: permission.value,
    values: permission.values,
  };
}

function mapOcapiPermissions(ocapi: BmRolePermissions): RolePermissionsInfo {
  // OCAPI uses snake_case for innermost permission fields (locale_id, etc.)
  // while SCAPI uses camelCase (localeId). Convert at this boundary.
  const result: Record<string, unknown> = {};
  if (ocapi.module) {
    result.module = {
      organization: (ocapi.module.organization ?? []).map(mapOcapiModulePermission),
      site: (ocapi.module.site ?? []).map(mapOcapiModulePermission),
    };
  }
  if (ocapi.functional) {
    result.functional = {
      organization: (ocapi.functional.organization ?? []).map(mapOcapiFunctionalPermission),
      site: (ocapi.functional.site ?? []).map(mapOcapiFunctionalPermission),
    };
  }
  if (ocapi.locale) {
    result.locale = {
      unscoped: (ocapi.locale.unscoped ?? []).map(mapOcapiLocalePermission),
    };
  }
  if (ocapi.webdav) {
    result.webdav = {
      unscoped: (ocapi.webdav.unscoped ?? []).map(mapOcapiWebdavPermission),
    };
  }
  return result as RolePermissionsInfo;
}

function mapScapiPermissionsToOcapi(perms: RolePermissionsInfo): BmRolePermissions {
  // Reverse: camelCase → snake_case for the inner locale field.
  const result: Record<string, unknown> = {};
  if (perms.module) {
    result.module = {
      organization: (perms.module.organization ?? []).map((permission) => ({...permission})),
      site: (perms.module.site ?? []).map((permission) => ({...permission})),
    };
  }
  if (perms.functional) {
    result.functional = {
      organization: (perms.functional.organization ?? []).map((permission) => ({...permission})),
      site: (perms.functional.site ?? []).map((permission) => ({...permission})),
    };
  }
  if (perms.locale) {
    result.locale = {
      unscoped: (perms.locale.unscoped ?? []).map((permission) => ({
        locale_id: permission.localeId,
        type: permission.type,
        value: permission.value,
        values: permission.values,
      })),
    };
  }
  if (perms.webdav) {
    result.webdav = {
      unscoped: (perms.webdav.unscoped ?? []).map((permission) => ({...permission})),
    };
  }
  return result as BmRolePermissions;
}

export class OcapiRolesBackend implements RolesBackend {
  readonly name = 'ocapi' as const;

  constructor(private instance: B2CInstance) {}

  async listRoles(options: ListRolesOptions = {}): Promise<ListRolesResult> {
    const result = await ocapiListBmRoles(this.instance, {start: options.start, count: options.count});
    const items = (result.data ?? []) as BmRole[];
    return {
      total: result.total ?? 0,
      start: result.start ?? 0,
      count: result.count ?? items.length,
      hits: items.map(mapOcapiRole),
    };
  }

  async getRole(roleId: string, options?: {expand?: ('users' | 'permissions')[]}): Promise<RoleInfo> {
    const role = await ocapiGetBmRole(this.instance, roleId, {expand: options?.expand});
    return mapOcapiRole(role);
  }

  async createRole(roleId: string, input?: CreateRoleInput): Promise<RoleInfo> {
    const role = await ocapiCreateBmRole(this.instance, roleId, {description: input?.description});
    return mapOcapiRole(role);
  }

  async deleteRole(roleId: string): Promise<void> {
    await ocapiDeleteBmRole(this.instance, roleId);
  }

  async getPermissions(roleId: string): Promise<RolePermissionsInfo> {
    const perms = await ocapiGetBmRolePermissions(this.instance, roleId);
    return mapOcapiPermissions(perms);
  }

  async setPermissions(roleId: string, permissions: RolePermissionsInfo): Promise<RolePermissionsInfo> {
    const updated = await ocapiSetBmRolePermissions(this.instance, roleId, mapScapiPermissionsToOcapi(permissions));
    return mapOcapiPermissions(updated);
  }

  async grantRole(roleId: string, login: string): Promise<void> {
    await ocapiGrantBmRole(this.instance, roleId, login);
  }

  async revokeRole(roleId: string, login: string): Promise<void> {
    await ocapiRevokeBmRole(this.instance, roleId, login);
  }
}

File: python/src/b2c_tooling_sdk/operations/bm_roles/ocapi_backend.py
# Copyright (c) 2025, Salesforce, Inc.
# SPDX-License-Identifier: Apache-2.0
# For full license text, see the LICENSE file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
"""OCAPI implementation of the Business Manager roles backend.

Mirrors ``src/operations/bm-roles/ocapi-backend.ts``. OCAPI uses snake_case for
some innermost permission fields (``locale_id``) while the canonical/SCAPI shape
uses camelCase (``localeId``); this module converts at that boundary.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal, cast

from b2c_tooling_sdk.operations.bm_roles import roles as _roles
from b2c_tooling_sdk.operations.bm_roles.roles import GetBmRoleOptions, ListBmRolesOptions
from b2c_tooling_sdk.operations.bm_roles.types import (
    CreateRoleInput,
    ListRolesOptions,
    ListRolesResult,
    RoleExpand,
    RoleInfo,
    RolePermissionsInfo,
)

if TYPE_CHECKING:
    from b2c_tooling_sdk.instance import B2CInstance


def _map_module_permission(permission: dict[str, Any]) -> dict[str, Any]:
    return {
        "application": permission.get("application"),
        "name": permission.get("name"),
        "type": permission.get("type"),
        "system": permission.get("system"),
        "value": permission.get("value"),
        "values": permission.get("values"),
    }


def _map_functional_permission(permission: dict[str, Any]) -> dict[str, Any]:
    return {
        "name": permission.get("name"),
        "type": permission.get("type"),
        "value": permission.get("value"),
        "values": permission.get("values"),
    }


def _map_locale_permission(permission: dict[str, Any]) -> dict[str, Any]:
    return {
        "localeId": permission.get("locale_id"),
        "type": permission.get("type"),
        "value": permission.get("value"),
        "values": permission.get("values"),
    }


def _map_webdav_permission(permission: dict[str, Any]) -> dict[str, Any]:
    return {
        "folder": permission.get("folder"),
        "type": permission.get("type"),
        "value": permission.get("value"),
        "values": permission.get("values"),
    }


def _map_ocapi_permissions(ocapi: dict[str, Any]) -> RolePermissionsInfo:
    """OCAPI (snake_case innermost fields) -> canonical (camelCase) permissions shape."""
    result: dict[str, Any] = {}

    module = ocapi.get("module")
    if module:
        result["module"] = {
            "organization": [_map_module_permission(p) for p in (module.get("organization") or [])],
            "site": [_map_module_permission(p) for p in (module.get("site") or [])],
        }

    functional = ocapi.get("functional")
    if functional:
        result["functional"] = {
            "organization": [_map_functional_permission(p) for p in (functional.get("organization") or [])],
            "site": [_map_functional_permission(p) for p in (functional.get("site") or [])],
        }

    locale = ocapi.get("locale")
    if locale:
        result["locale"] = {"unscoped": [_map_locale_permission(p) for p in (locale.get("unscoped") or [])]}

    webdav = ocapi.get("webdav")
    if webdav:
        result["webdav"] = {"unscoped": [_map_webdav_permission(p) for p in (webdav.get("unscoped") or [])]}

    return result


def _map_scapi_permissions_to_ocapi(perms: RolePermissionsInfo) -> dict[str, Any]:
    """Reverse: canonical (camelCase) -> OCAPI (snake_case innermost ``locale_id``)."""
    result: dict[str, Any] = {}

    module = perms.get("module")
    if module:
        result["module"] = {
            "organization": [dict(p) for p in (module.get("organization") or [])],
            "site": [dict(p) for p in (module.get("site") or [])],
        }

    functional = perms.get("functional")
    if functional:
        result["functional"] = {
            "organization": [dict(p) for p in (functional.get("organization") or [])],
            "site": [dict(p) for p in (functional.get("site") or [])],
        }

    locale = perms.get("locale")
    if locale:
        result["locale"] = {
            "unscoped": [
                {
                    "locale_id": p.get("localeId"),
                    "type": p.get("type"),
                    "value": p.get("value"),
                    "values": p.get("values"),
                }
                for p in (locale.get("unscoped") or [])
            ]
        }

    webdav = perms.get("webdav")
    if webdav:
        result["webdav"] = {"unscoped": [dict(p) for p in (webdav.get("unscoped") or [])]}

    return result


def _map_ocapi_role(ocapi: dict[str, Any]) -> RoleInfo:
    permissions = ocapi.get("permissions")
    return RoleInfo(
        id=ocapi.get("id") or "",
        description=ocapi.get("description"),
        user_count=ocapi.get("user_count"),
        user_manager=ocapi.get("user_manager"),
        # OCAPI returns permissions inline on the role when expanded, same as SCAPI.
        # Map snake_case -> camelCase to match the canonical RoleInfo shape so that
        # callers see consistent data after a fallback from SCAPI to OCAPI.
        permissions=_map_ocapi_permissions(permissions) if permissions else None,
        raw=ocapi,
    )


class OcapiRolesBackend:
    """Business Manager roles backend implemented against the OCAPI Data API."""

    def __init__(self, instance: B2CInstance) -> None:
        self._instance = instance

    @property
    def name(self) -> Literal["ocapi"]:
        return "ocapi"

    async def list_roles(self, options: ListRolesOptions | None = None) -> ListRolesResult:
        opts = options or ListRolesOptions()
        result = await _roles.list_bm_roles(self._instance, ListBmRolesOptions(start=opts.start, count=opts.count))
        items = cast("list[dict[str, Any]]", result.get("data") or [])
        count = result.get("count")
        return ListRolesResult(
            total=result.get("total") or 0,
            start=result.get("start") or 0,
            count=count if count is not None else len(items),
            hits=[_map_ocapi_role(item) for item in items],
        )

    async def get_role(self, role_id: str, *, expand: list[RoleExpand] | None = None) -> RoleInfo:
        role = await _roles.get_bm_role(self._instance, role_id, GetBmRoleOptions(expand=expand))
        return _map_ocapi_role(role)

    async def create_role(self, role_id: str, role_input: CreateRoleInput | None = None) -> RoleInfo:
        role = await _roles.create_bm_role(
            self._instance, role_id, description=role_input.description if role_input else None
        )
        return _map_ocapi_role(role)

    async def delete_role(self, role_id: str) -> None:
        await _roles.delete_bm_role(self._instance, role_id)

    async def get_permissions(self, role_id: str) -> RolePermissionsInfo:
        permissions = await _roles.get_bm_role_permissions(self._instance, role_id)
        return _map_ocapi_permissions(permissions)

    async def set_permissions(self, role_id: str, permissions: RolePermissionsInfo) -> RolePermissionsInfo:
        updated = await _roles.set_bm_role_permissions(
            self._instance, role_id, _map_scapi_permissions_to_ocapi(permissions)
        )
        return _map_ocapi_permissions(updated)

    async def grant_role(self, role_id: str, login: str) -> None:
        await _roles.grant_bm_role(self._instance, role_id, login)

    async def revoke_role(self, role_id: str, login: str) -> None:
        await _roles.revoke_bm_role(self._instance, role_id, login)


__all__ = ["OcapiRolesBackend"]
